### ISSUE TITLE
User (e.g. health care worker) can not differentiate the app notifications to unprotected or protected periods

### ISSUE THREAD

**Issue by rvbaer** [corona-warn-app/cwa-wishlist#251]
Had this feature discussed in Personal contact diary #196 
Recommendation in #196  was to open a new feature request

## Feature description

The app should display the time or time slots for each critical contact that has happened. 
Not just by days and also not only the last contact as now but preferably by hours for each contact. 

If this causes data protection topics, longer periods, such as 3-hour periods, would still be better than the current solution.

## Problem and motivation

The app is currently unusable for people who are in frequent professional contact with potentially or real infected people (eg health workers). These people are well protected during professional activities and have a low risk. 

On the other hand, they may have critical contacts in their private sphere without apropriate protection.
They might even meet with people from their work environment in diifferent scenarios - protected (at work) and unprotected (in private).

The user can currently not differentiate the notifications of the app as shown below to unprotected or protected periods. Accordingly, theses user will probalbly ignore indications of an increased risk as they know they had for sure enconters at work.

If the user receives the time of critical contacts he will often know how his protection status at that time was (work hours vs. private meetings) and can act accordingly responsible.  Thus he will protect others if there is a serious danger but continues to work if he was protected (and continues to help others by careing for COVID patients).

## Is this something you're interested in working on

Yes  

![Screenshot_20201028-090152_Corona-Warn](https://user-images.githubusercontent.com/73844115/98044414-1f24df00-1e27-11eb-8742-065d498bbc35.jpg)

**Comment by ndegendogo** [comment]
@rvbaer afaik, the current RKI recommendation for health workers is to disable bluetooth during their work time / while they wear protection gear. Just wanted to point this out, in case you didn't know.

Of course, it is a manual procedure and error-prone (you can easily forget to switch off, or to switch on after work again).

**Comment by MikeMcC399** [comment]
@ndegendogo & @rvbaer 

The current RKI recommendations for health workers (in German) are in the [RKI Corona-Warn-App FAQs](https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/WarnApp/Warn_App.html).

"Wie soll die Corona-Warn-App vom Personal von Einrichtungen der Gesundheitsversorgung bzw. vom Personal bei Tätigkeiten wie Krankentransporte verwendet werden?"

From https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/WarnApp/Warn_App.html scroll down to section

**Epidemiologische Fragen**
then read:
"Wie soll die Corona-Warn-App vom Personal von Einrichtungen der Gesundheitsversorgung bzw. vom Personal bei Tätigkeiten wie Krankentransporte verwendet werden?"

Depending on the situation, RKI sometimes recommends deactivating Bluetooth, in other situations they recommend disabling it.

I don't know why they talk about disabling Bluetooth by the way (which affects other apps using Bluetooth), instead of directly stopping Exposure Logging in CWA.

**Comment by rvbaer** [comment]
@MikeMcC399
Agree fully with you - > I cannot deactivate bluetooth as it is required for more things than the CWA.
Even at work a smart watch and headsets are in some setups required (e.g. Emergency Medicine)

@ndegendogo 
Thanks for the comment but as you nicely mention it is not helpful at all to ask health care workers to switch CWA on and off.
Many health care workers have continuously changing situations from treating potential infected patients (with FFP2 mask), to drinking a cup of tea with some colleagues in the same room to walking down a corridor with a simple mask only. 

Should all of them continuously switch their phones on and off? This is not feasible at all!!

I have for myself started using the workaround shared by @daimpi here
https://github.com/corona-warn-app/cwa-wishlist/issues/196#issuecomment-720684343

It is working excellent and I can only recommend all to install it.
The Warn-App-Companion gives me even the exact map location and the time of a contact.
Had so far one contact (Übertragungsrisiko 6 = very low) and could easy remember that I was in an ambulance car at that time.  And the contact person was for sure outside.
The information provided by Warn-App-Companion is anonymous and thus in line with data security laws. 
I c strongly recommend to RKI and the developers to integrate the information provided by Warn-App-Companion into future versions of the CWA app. This would certainly allow many people in the health business but also others to understand better where they had been at risk and how well the been protected at this time. 
It would certainly increase the acceptance of personal “volunteer” quarantine   measures if people see where they had been exposed to risk and for how long. Especially now where more and more people start wearing FFP2 mask in public.

**Comment by tangomanni** [comment]
@rvbaer 
Since 3 days I have a warning about a low risk contact. As I want to know on which day this contact happened I installed the warn-app-companion and ramble . But now  I don't know how to proceed ? Where is the database which should be exported by ramble ?

**Comment by Ein-Tim** [comment]
@rvbaer 

> I strongly recommend to RKI and the developers to integrate the information provided by Warn-App-Companion into future versions of the CWA app. This would certainly allow many people in the health business but also others to understand better where they had been at risk and how well the been protected at this time.


So sorry to say that but RKI already denied to show anymore Information about Risk Encounters in CWA, see this comment: https://github.com/corona-warn-app/cwa-wishlist/issues/100#issuecomment-690279530

Although, here in the wishlist Repo still many Issues are open which request to show at least some more information, f.e.: #178,

**Comment by daimpi** [comment]
@Ein-Tim @rvbaer 
> > I strongly recommend to RKI and the developers to integrate the information provided by Warn-App-Companion into future versions of the CWA app. This would certainly allow many people in the health business but also others to understand better where they had been at risk and how well the been protected at this time.
> 
> So sorry to say that but RKI already denied to show anymore Information about Risk Encounters in CWA, see this comment: [#100 (comment)](https://github.com/corona-warn-app/cwa-wishlist/issues/100#issuecomment-690279530)

I know that this statement by RKI but tbh I don't see it as stating "we will not provide any more info on encounters" but rather as "we will not show more than two colors" which are two different things hence #100 has been wrongfully closed imho.

But apart from that: The request stated in OP
> Not just by days and also not only the last contact as now but preferably by hours for each contact.

Would require Google/Apple to change their Exposure Notification Framework (ENF) b/c exact time data is currently not exposed to CWA (see #206 for more discussion on this topic).

@tangomanni I agree with @Ein-Tim but you can also join us on the [community Slack](https://github.com/corona-warn-app/cwa-documentation/issues/379) where we have a dedicated channel for the companion-app 🙂.

**Comment by tangomanni** [comment]
@daimpi I will join this community slack. Thank you.

**Comment by rvbaer** [comment]
@tangomanni

**RaMBLE & Warn-App-Companion manual**

First  you have to manually export the database in RaMBLE **each time** you want to check your exposure. This is in Top Line Menue "Export Database".
After you have done so you start Warn-App-Companion. In this APP you have to manually set the Top Menue to "RaMBLE Modus" **each time** you want to check.  It then synchronizes you Exposure from RaMBLE with the "Diaganoseschlüssel" data from CWA. 

Matches are highlighted in the third line. If you touch them details will be shown

![Screenshot_20201108-183931_Warn-App-Companion](https://user-images.githubusercontent.com/73844115/98478220-15202900-21f7-11eb-8c8b-acb769c09104.jpg)

> 
> @rvbaer
> Since 3 days I have a warning about a low risk contact. As I want to know on which day this contact happened I installed the warn-app-companion and ramble . But now I don't know how to proceed ? Where is the database which should be exported by ramble ?

**Comment by tangomanni** [comment]
Thank you for all the support from different people. I appreciate that.
I have started ramble and I got two contacts for today and no diagnosis keys. I wonder that I got these contacts although I live alone and didn't go out from my flat today. Maybe blootooth goes through walls or a lot of meters when balcony door is open.

**Comment by ndegendogo** [comment]
> Maybe blootooth goes through walls or a lot of meters when balcony door is open.

Yes, I guess this is true. cwa compagnion picks up all signals it sees, however weak or short.

**Comment by mh-** [comment]
And @tangomanni it's normal that there are no Diagnosis Keys for today published yet. Tomorrow you will see some.

**Comment by MikeMcC399** [comment]
@tangomanni 
> Maybe blootooth goes through walls or a lot of meters when balcony door is open.

I was really surprised when I ran [Beacon Scope](https://play.google.com/store/apps/details?id=com.davidgyoungtech.beaconscanner&hl=en) just how many Bluetooth beacons for Exposure Notification were close to me. I live in a flat with strong concrete walls and I thought that I would be shielded from other users, but I guess the signal goes through windows and maybe even concrete.

**Comment by tangomanni** [comment]
> @tangomanni
> 
> > Maybe blootooth goes through walls or a lot of meters when balcony door is open.
> 
> I was really surprised when I ran [Beacon Scope](https://play.google.com/store/apps/details?id=com.davidgyoungtech.beaconscanner&hl=en) just how many Bluetooth beacons for Exposure Notification were close to me. I live in a flat with strong concrete walls and I thought that I would be shielded from other users, but I guess the signal goes through windows and maybe even concrete.

all doors closed -> ramble shows 20 new devices (3 in range)...wooww.. sometimes i think one should not know everything

**Comment by mh-** [comment]
Yes, and also note that each device changes the transmitted data every 10..20 minutes, and each time _a new encounter starts_ because at this time nobody knows if it's the same device, or another device. So you will probably see many more "encounters" than actual people.
"3 in range" means that they are actually there, but the other 17 could just be old versions of these 3.

**Comment by rvbaer** [comment]
> 
> 
> @Ein-Tim @rvbaer
> 
> > > I strongly recommend to RKI and the developers to integrate the information provided by Warn-App-Companion into future versions of the CWA app. This would certainly allow many people in the health business but also others to understand better where they had been at risk and how well the been protected at this time.
> > 
> > 
> > So sorry to say that but RKI already denied to show anymore Information about Risk Encounters in CWA, see this comment: [#100 (comment)](https://github.com/corona-warn-app/cwa-wishlist/issues/100#issuecomment-690279530)
> 
> I know that this statement by RKI but tbh I don't see it as stating "we will not provide any more info on encounters" but rather as "we will not show more than two colors" which are two different things hence #100 has been wrongfully closed imho.
> 
> But apart from that: The request stated in OP
> 
> > Not just by days and also not only the last contact as now but preferably by hours for each contact.
> 
> Would require Google/Apple to change their Exposure Notification Framework (ENF) b/c exact time data is currently not exposed to CWA (see #206 for more discussion on this topic).
> 


@Ein-Tim & @daimpi
Thanks for pointing out to #100

When #100 was closed by RKI we had around 1900 cases per day. 
At that time it was  correct to have only limited information as likelihood of exposure was low and capacity of laboratories was high. In addition hardly anybody wore FFP2 mask. 
Now we have more than ten time as much cases and the ratio exposure / lab capacity has changed totally from 10th September. And walking through a supermarket around 30% of shoppers wear FFP2 mask

**RKI** has to consider to give out more information to prevent user from either **ignoring the information** or **overflowing the Lab test capacities**.

>Would require Google/Apple to change their Exposure Notification Framework (ENF) b/c exact time data is currently not exposed to CWA (see #206 for more discussion on this topic).

In addition as it works with the two apps RaMBLE & Warn-App-Companion at the moment nicely it will be easy to setup something similar but more convenient without Google/Apple exposing more data


![2020-11-08 19_28_08-corona infektion deutscchland - Google Suche](https://user-images.githubusercontent.com/73844115/98481162-a93ec000-21f8-11eb-815f-7efa31a5c7ab.png)

**Comment by daimpi** [comment]
@rvbaer I agree that it would be nice to have the warn-app-companion (WAC) functionality provided by CWA. But this is not possible currently b/c Bluetooth access is not allowed for ENF apps by Google/Apple. See [this thread](https://github.com/corona-warn-app/cwa-wishlist/issues/100) for further discussion on this topic. But even if CWA could access BT: this could only work on Android devices as iOS is filtering out all the BLE signals containing rolling proximity identifiers (RPIs) already on the OS level. So nothing except for the Exposure Notification Framework (ENF) which is directly integrated into iOS can even see them.

And even under Android this hypothetical "direct CWA recording" would probably end up being quite unreliable, as many manufacturers implement "energy optimizations" which [kill](https://dontkillmyapp.com/) background apps all the time.

**Comment by rvbaer** [comment]
@daimpi 
Thanks for the background on why something is possible with warn-app-companion (WAC) & RaMBLE but not with an integrated App. 

I understand that pressing the "export database" button in RaMPLE and the import button in WAC is okay in Android. But for Google it is not okay not have an integrated App with the same functionalty. 
Sounds strange but obviously these are Google's rules - which are as a consequence also strange.

I am still convinced that it is worth the effort to fight for a better app which supports the users by giving time and location of critical enncounters. Such an improved App would
A) reduce the number of people who want a test without a need 
B) reduce the number of people who do **not** bring themself in quarantaine after the App alarmed them
Such a feature would thus simply increase the "quality"

Especially now with the high exposure risk the RKI should join and also fight for this improvement.

**Comment by Blackjacx** [comment]
All comments here deal with a feature an app that only works for Android. Please do not forget the iOS users!

Today I posted the comment https://github.com/corona-warn-app/cwa-wishlist/issues/234#issuecomment-723671593 which explains how to help more health staff to use the CWA app since many of them simply don't use it because it would be permanently red. Isn't it better if they use it everywhere except in hospital instead of not using it at all?

I think until Apple and Google are not convinced reporting exact location and time information we should implement possible solutions that help health staff to get on this app. Due to privacy reasons, the integration of location/time info can still be far in the future.

**Comment by ndegendogo** [comment]
> Please do not forget the iOS users!

YES!

> ... automatically using geofencing ...

well ... currently not possible, due to restrictions by Apple and Google ... an app that uses their ENF framework is (currently) not allowed to use location services ...

But this is the wishlist

**Comment by Blackjacx** [comment]
> well ... currently not possible, due to restrictions by Apple and Google ... an app that uses their ENF framework is (currently) not allowed to use location services ...
> 
> But this is the wishlist

Oh no, I didn't know that point. Actually it was also not so easy to find on the interne. But for any curious guy like me, it is stated under ["How will the system protect user privacy and security?"](https://covid19-static.cdn-apple.com/applications/covid19/current/static/contact-tracing/pdf/ExposureNotification-FAQv1.2.pdf)

**Comment by ndegendogo** [comment]
@Blackjacx yes. And also in section 10 of same document:

> 10. How will apps get approval to use this system?
[...]There will be restrictions on the data that apps can collect when using the API, including not being able to request access to location services ...

**Comment by Blackjacx** [comment]
Awful 😅

**Comment by alanrick** [comment]
The Irish App also uses the Apple/Google services and includes this capability. Why not sync with their team to discover how difficult it is to implement and what the benefits/side-effects are in practice?
![72AEDFE5-6779-44A0-8496-260C3F64FDE3](https://user-images.githubusercontent.com/7034358/98996839-d6121080-2533-11eb-8e7a-d4e58a9af7cb.jpeg)

**Comment by stherold** [comment]
Nice this works without geofencing so it must be accespted by the big players.

**Comment by alanrick** [comment]
Philosophical...Most important for me would be the collaboration between the different EU app development groups. More experience-based decisions , less arrogance, more transparency.

**Comment by Brazier85** [comment]
we need this feature as soon as possible. I just removed the app from my phone because of the new way the risk is calculated. I am working at a local testing site. So I have a lot of contacts to people that are infected. Till last week no one of our workers had a red app. Just 6 to 9 low risk contacts. Now there are a few red apps - we need to have a way to disable the tracking in certain locations or time periods...

**Comment by mh-** [comment]
Did you consider switching Bluetooth off during these periods where you are protected by special masks, clothing etc?

**Comment by Ein-Tim** [comment]
If you don't want to switch off Bluetooth (for whatever reason) you could also only switch off "Risk Detection" in the App. I know that this is not the best option, but I think it's still better than to deinstall the whole App.

**Comment by MikeMcC399** [comment]
The [RKI](https://www.rki.de/EN/) recommendations which I referred to in https://github.com/corona-warn-app/cwa-wishlist/issues/251#issuecomment-723553121 have now moved.

To find them, 
- go to [**Infektionsketten digital unterbrechen mit der Corona-Warn-App**](https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/WarnApp/Warn_App.html)
- scroll down to section "**Fragen zum öffentlichen Gesundheitsdienst/ Labor/ Medizinisches Personal (20.11.2020)**"
- click on "**Wie die Corona-Warn-App vom medizinischen Personal (inklusive Krankentransporte) verwendet werden soll**"

**Comment by alanrick** [comment]
You can also easily switch it off in the phone settings. Nevertheless, I do like the reminder function that the Irish and NHS Apps provide.  
So if you need the reminder function it might be worth using the Irish app instead because it's compatible. The NHS isn't.

![EpTpvwvWMAEUtpm](https://user-images.githubusercontent.com/7034358/102781522-187f0680-4398-11eb-8006-3bb0db4261d3.jpeg)

**Comment by alanrick** [comment]
> The [RKI](https://www.rki.de/EN/) recommendations which I referred to....

I much prefer @Ein-Tim's suggestion to switching off bluetooth. Switching off bluetooth deactivates all blue-tooth devices such as key-finders, locks, headphones...

**Comment by MikeMcC399** [comment]
@alanrick 
Each of the two options has its own disadvantages:

- if you switch off Bluetooth you may lose other functionality you need, like headphones as you mention
- if you disable Exposure Logging, you not only disable the exchange of Bluetooth signals, you also disable exposure checks, which, with a continuous WLAN connection, will be triggered approximately every 4 hours, so you might miss out on a timely high risk warning from the app

**Comment by heinezen** [comment]
This issue has been somewat mitigated by two changes:

* The exact data of a high risk contact is now sown in the app
* Users can manually enter work locations and contacts in the app

Differentiating between protected and unprotected periods has not been addressed yet.

----
Corona-Warn-App Open Source Team

**Comment by stherold** [comment]
What is the app doing with these information: `work locations` and `contacts in the app`

**Comment by Blackjacx** [comment]
What is the app doing with these information: `work locations` and `contacts in the app`

**Comment by heinezen** [comment]
@stherold This information can be entered into the app's contact diary. The app itself does nothing with it, except storing the info. If users are tested positive or see a high risk contact in the app, they can use the contact diary to check where they were and who they met if they don't remember.

---
Corona-Warn-App Open Source Team

**Comment by alanrick** [comment]
@rvbaer Yesterday's update to the CoronaMelder app (Netherlands) added a pause function so that protected medical workers don't get false-positive notifications. This app is available in German and is compatible with the CWA. 

If  you are a medical worker it might be sensible to use CoronaMelder until the CWA offers the pause function, switching to the CWA if you test positive (or are tested by a lab that supports the CWA).

![Datei_001](https://user-images.githubusercontent.com/7034358/106247899-7e20a700-6210-11eb-8194-072eb26bfdce.jpeg)
![Datei_000](https://user-images.githubusercontent.com/7034358/106247913-824cc480-6210-11eb-81c0-e35dfbba235e.png)
![Datei_002](https://user-images.githubusercontent.com/7034358/106247883-78c35c80-6210-11eb-828f-94f7b6af50b7.jpeg)

**Comment by ndegendogo** [comment]
@alanrick wow - cool feature!

**Comment by ndegendogo** [comment]
@dsarkar for iPhone, the method [setExposureNotificationEnabled](https://developer.apple.com/documentation/exposurenotification/enmanager/3583729-setexposurenotificationenabled) is a good candidate for this feature.

> Using this method to disable exposure notification stops scanning and Bluetooth advertising, but diagnosis keys and data remain.

**Comment by dsarkar** [comment]
@ndegendogo, @alanrick, @rvbaer 

We have updated the internal ticket with your latest contributions. Thanks.

**Comment by Ein-Tim** [comment]
This is closely related to #108. I guess either #108 should be merged with this issue or vice versa.

**Comment by MarcusZuber** [comment]
Maybe the possibility to exclude defined time-slots from the risk calculation, where the user feals confident, that he was was well protected, would be a possibility. This could maybe done with the help of the contact-log ("Kontakt-Tagebuch"). 

If the user can only set these "safe time-slots" for a short period (maybe for one day) , it is not possible to use this feature to exactly get the time of the contact (for data privacy reasons).

This would help to understand such warnings like "High risk contacts on <n> days, last day X", where the user nows that on X he was always protected and maybe would have noted this in his logbook.

**Comment by ndegendogo** [comment]
@MarcusZuber agreed. But a useful "short" safe time slot would be rather 1 hour than one whole day.

I am riding to work and back home with public transport. During this time I wear my FFP2 mask.
At lunch time I go into the lunchroom at our office. During this time I don't wear any mask.

**Comment by MarcusZuber** [comment]
Ah sorry, I meant, it is only possible for one day to edit arbitrary "safe time slots" and not, that the whole day is marked as safe.
